### ISSUE TITLE
Whether it is normal that the amount of communication data is huge when I do parallel on multi-machine？

### ISSUE THREAD

**Issue by ljz756245026** [facebookresearch/fairscale#797]
## ❓ Questions and Help
I use fairscale optimizer state sharding（OSS） to train on multi-machine， and I found that the amount of communication data is huge， which is around 2x than that of distribute data parallel. Whether it is normal?

When I do parallel on multi-machine by DDP(distribute data parallel), the amount of communication data is as follows:
![image](https://user-images.githubusercontent.com/21344631/133364492-b46dc4a3-62cd-463e-b6d6-50a7b904a5a5.png)
When I do parallel on multi-machine by OSS(fairscale optimizer state sharding), the amount of communication data is as follows:
![image](https://user-images.githubusercontent.com/21344631/133364555-4a167158-8bcd-4c45-8a7e-b5a1166a37d5.png)

All the image above is from master server.
I found that the amount of communication of OSS is around double than that of DDP.
Furthermore, I found that the slave server would send more data than master server, is there any website or paper can explain the phenomena?

**Comment by min-xu-ai** [comment]
Are you using adam or SGD? That change the volume too. I haven't done a detailed calculation but @blefaudeux may have it.

**Comment by blefaudeux** [comment]
- raw coms: yes it's expected, in that with DDP all the gradients are moved over (roughly the size of the model), and OSS adds to that the model shards.

- I'm not sure what you mean by "slave" and "master" server, but if this refers to the main rank on which the checkpoint is consolidated, then yes this is expected given the fact that all the other ranks will send their optimizer state over there.

**Comment by blefaudeux** [comment]
To add to what @min-xu-ai wrote above, using Adam or SGD will change the volume of data exchanged when checkpointing, but not the volume of data used while training.

**Comment by ljz756245026** [comment]
> Are you using adam or SGD? That change the volume too. I haven't done a detailed calculation but @blefaudeux may have it.

yes，I use SGD to train the model. The model size is around 96 MB. However, I found that the volume of data, which is around 189 / 264GB(OSS),  is so huge.

**Comment by ljz756245026** [comment]
> * raw coms: yes it's expected, in that with DDP all the gradients are moved over (roughly the size of the model), and OSS adds to that the model shards.
> * I'm not sure what you mean by "slave" and "master" server, but if this refers to the main rank on which the checkpoint is consolidated, then yes this is expected given the fact that all the other ranks will send their optimizer state over there.

I use ```tcp``` as the init_method in ```dist.init_process_group```. The master are the IP address I used in init_method. Other machine(s) are slave machine. 
```
 dist.init_process_group(backend=args.dist_backend, init_method=args.dist_url,
                                world_size=args.world_size, rank=args.rank)
```

**Comment by blefaudeux** [comment]
> > Are you using adam or SGD? That change the volume too. I haven't done a detailed calculation but @blefaudeux may have it.
> 
> yes，I use SGD to train the model. The model size is around 96 MB. However, I found that the volume of data, which is around 189 / 264GB(OSS), is so huge.

the amount of comms will depend on the number of GPUs in the training, the number of steps you do, and whether you consolidate a checkpoint often or not. Twice DDP is about right, no big surprise here. 

If you're training with AMP though, you could use the "broadcast_fp16" option in OSS, it should remove about 1/4th of the communication volume (be sure to use current master though, this option was broken until recently)

**Comment by min-xu-ai** [comment]
> > Are you using adam or SGD? That change the volume too. I haven't done a detailed calculation but @blefaudeux may have it.
> 
> yes，I use SGD to train the model. The model size is around 96 MB. However, I found that the volume of data, which is around 189 / 264GB(OSS), is so huge.

Did you just run only 1 iteration of forward and backward passes? If so, 189GB is huge. However, if you run many iterations, then, every iteration the transfer size 96MB times the number of iterations, you can get to 189GB for sure.

**Comment by ljz756245026** [comment]
> > > Are you using adam or SGD? That change the volume too. I haven't done a detailed calculation but @blefaudeux may have it.
> > 
> > 
> > yes，I use SGD to train the model. The model size is around 96 MB. However, I found that the volume of data, which is around 189 / 264GB(OSS), is so huge.
> 
> the amount of comms will depend on the number of GPUs in the training, the number of steps you do, and whether you consolidate a checkpoint often or not. Twice DDP is about right, no big surprise here.
> 
> If you're training with AMP though, you could use the "broadcast_fp16" option in OSS, it should remove about 1/4th of the communication volume (be sure to use current master though, this option was broken until recently)

**Comment by ljz756245026** [comment]
> > > Are you using adam or SGD? That change the volume too. I haven't done a detailed calculation but @blefaudeux may have it.
> > 
> > 
> > yes，I use SGD to train the model. The model size is around 96 MB. However, I found that the volume of data, which is around 189 / 264GB(OSS), is so huge.
> 
> Did you just run only 1 iteration of forward and backward passes? If so, 189GB is huge. However, if you run many iterations, then, every iteration the transfer size 96MB times the number of iterations, you can get to 189GB for sure.

I train 5 epochs and there are around 300 batches in 1 epochs.What is interesting is that for OSS, the volume of TX data and RX are different, which makes me confused. In DDP, the volume of TX and RX data are the same. Why it is different in OSS method. Did you add some tricks on communication?

**Comment by min-xu-ai** [comment]
In that case, 189/264GB makes sense.

In terms of OSS's tx/rx, please check the oss.py code. It splits optimizer states unevenly between master and slave so it is expected the tx/rx being different.

**Comment by ljz756245026** [comment]
> In that case, 189/264GB makes sense.
> 
> In terms of OSS's tx/rx, please check the oss.py code. It splits optimizer states unevenly between master and slave so it is expected the tx/rx being different.

Thank you for your reply and patient explanation! Have a nice day！ 🌞

**Comment by blefaudeux** [comment]
>  In DDP, the volume of TX and RX data are the same. Why it is different in OSS method. Did you add some tricks on communication?

I tried to explain that above, but it also depends on your calls to `.consolidate_state_dict()`, if any (I wouldn't recommend more than once an hour basically, but tastes and colours..). By default this will consolidate the state on the reference rank, meaning that this rank will receive more, and the others will transmit more. 

Another explanation is pointed by Min above, the model shards follow the tensor boundaries, so they will not be strictly equivalent, but empirically the difference should not be much.

DDP does not shard, all the ranks do strictly the same thing (except for a different data chunk hopefully), so both do not happen.
